### ISSUE TITLE
Details of the Circular Convolution

### ISSUE THREAD

**Issue by LetsGoFir** [zju3dv/snake#8]
The backbone is some 'CirConv's in your paper, that is
![image](https://user-images.githubusercontent.com/59643844/74497601-1c8cfd00-4f19-11ea-9260-6f012aeee67b.png)
Then is f_N the hstack(or vstack) of N features? If I want to do the forward, should I take the stack of N features as input, or you really do this by the formula of right side?

**Comment by LetsGoFir** [comment]
I want to implement it with https://github.com/kazuto1011/circular-conv-pytorch.

**Comment by pengsida** [comment]
Circular convolution:

```
class CircConv(nn.Module):
    def __init__(self, state_dim, out_state_dim=None, n_adj=4):
        super(CircConv, self).__init__()

        self.n_adj = n_adj
        out_state_dim = state_dim if out_state_dim is None else out_state_dim
        self.fc = nn.Conv1d(state_dim, out_state_dim, kernel_size=self.n_adj*2+1)

    def forward(self, input, adj):
        input = torch.cat([input[..., -self.n_adj:], input, input[..., :self.n_adj]], dim=2)
        return self.fc(input)
```

**Comment by LetsGoFir** [comment]
Thanks a lot! I think the input shape is Bx1xDxN, in which B, D, N are the batch size, feature dimension and the number of vertices respectively, right?

**Comment by pengsida** [comment]
Yes.

**Comment by LetsGoFir** [comment]
Thanks!

**Comment by LetsGoFir** [comment]
> Yes.

![image](https://user-images.githubusercontent.com/59643844/74519729-3ba78100-4f51-11ea-849d-4587dd2587be.png)
Hello, I wonder whether the first CircConvBlock has the skip connection?

**Comment by LetsGoFir** [comment]
![image](https://user-images.githubusercontent.com/59643844/74521794-f5ecb780-4f54-11ea-9167-642e68fbb011.png)
And it seems like that the output of last CirConv is sumed twice here, isn't it? Should I sum it once?

**Comment by pengsida** [comment]
> Hi, I have new questions here~
> ![image](https://user-images.githubusercontent.com/59643844/74519729-3ba78100-4f51-11ea-849d-4587dd2587be.png)
> 1.I wonder whether the first CircConvBlock has the skip connection?
> 2.Could you tell me the channels of every Conv layers and the Maxpool size?

1. The first CircConvBlock does not have a skip connection.
2. The channel number is 128.
3. The maxpool aggregates the features of all vertices.

**Comment by pengsida** [comment]
> ![image](https://user-images.githubusercontent.com/59643844/74521794-f5ecb780-4f54-11ea-9167-642e68fbb011.png)
> And it seems like that the output of last CirConv is sumed twice here, should I sum it once? Or the output(gary lines below) of every CirConv includes the skip connected tensor?

The output of last CircConv is concatenated with the features of previous CircConv.

**Comment by LetsGoFir** [comment]
> > ![image](https://user-images.githubusercontent.com/59643844/74521794-f5ecb780-4f54-11ea-9167-642e68fbb011.png)
> > And it seems like that the output of last CirConv is sumed twice here, should I sum it once? Or the output(gary lines below) of every CirConv includes the skip connected tensor?
> 
> The output of last CircConv is concatenated with the features of previous CircConvs.

Thanks again, and I will have a try.

**Comment by LetsGoFir** [comment]
@pengsida Hi~ I have 2 little questions here.
![image](https://user-images.githubusercontent.com/59643844/74604713-70573c00-50fb-11ea-9492-9f16547a09f4.png)
Do you sample N=128 vertices for every component box when handling multi-component objects, or the total number is 128? I found that there are many small boxes.
Another question, do you compute the loss separately for the component box when handling the multi-component objects?

**Comment by pengsida** [comment]
1. I sample N vertices for each component contour.
2. Yes, I compute the loss separately for each component.

**Comment by LetsGoFir** [comment]
> 1. I sample N vertices for each component contour.
> 2. Yes, I compute the loss separately for each component.

Thanks for your promptly reply!

**Comment by LetsGoFir** [comment]
> 1. I sample N vertices for each component contour.
> 2. Yes, I compute the loss separately for each component.

How do you treat the instances whose perimeter is less than 100? I want to drop them because there is no enough vertices to sample.

**Comment by pengsida** [comment]
I sample contour points using the code at https://github.com/zju3dv/snake/issues/3

**Comment by LetsGoFir** [comment]
> I sample contour points using the code at #3

Yeah, I am using your code to sample the coarse poly. Although I can get enough points by it, but there are many points too 'precise', like [119.6212, 158] and [119.5212, 158], and when I try to get the correspond feature, they are the same. How about you?

**Comment by pengsida** [comment]
I did not notice this problem.
I use grid_sample to extract the features.

**Comment by LetsGoFir** [comment]
> I did not notice this problem.
> I use grid_sample to extract the features.

Ok, then for each instance, do you upsample them to a fixed size? Like (224,224) or (512,512).

**Comment by pengsida** [comment]
No, I keep the original size.

**Comment by LetsGoFir** [comment]
> No, I keep the original size.
Hi~

It seems like you use the CenterNet to get the embedding of each vertex, so if it is true, do you fix the backbone(model released by CenterNet) and train only your deepsnake, or you train them all?

As for the dataloader, I noticed that the CenterNet convert the Pascal VOC to COCO format, did you do the same for SBD?

**Comment by pengsida** [comment]
1. I train the snake with CenterNet together.
2. Yes, I converted the SBD to COCO format.

**Comment by LetsGoFir** [comment]
> 1. I train the snake with CenterNet together.
> 2. Yes, I converted the SBD to COCO format.

Thanks for your reply!

**Comment by LetsGoFir** [comment]
> 1. I train the snake with CenterNet together.
> 2. Yes, I converted the SBD to COCO format.

Hi~ I noticed that you do inference with 3 iteration, and I want to know whether you train only 1 iteration when training?

**Comment by pengsida** [comment]
I train three iterations during training.

**Comment by LetsGoFir** [comment]
> I train three iterations during training.

Then how about the loss_iter when convergence?

**Comment by pengsida** [comment]
About 2.

**Comment by LetsGoFir** [comment]
> About 2.

Thanks, now I can converge to below 10 in the first epoch, training only on instances having only single component.

**Comment by LetsGoFir** [comment]
> About 2.

Hi~
How much time for an single epoch training? It takes me about 3.5 hours with 8 batchsize and 2GPU.

**Comment by pengsida** [comment]
An epoch takes about 5 minutes with batchsize 80 and 4 gpus on Sbd dataset.

**Comment by LetsGoFir** [comment]
> An epoch takes about 5 minutes with batchsize 80 and 4 gpus on Sbd dataset.

Fine, how about CityScapes?

**Comment by pengsida** [comment]
About 3 minutes on Cityscapes.
The loss is about 2.

**Comment by pengsida** [comment]
Our paper has been accepted, and we will release it soon.

**Comment by LetsGoFir** [comment]
> Our paper has been accepted, and we will release it soon.

Congratulations!

**Comment by LetsGoFir** [comment]
I am so happy to see you releasing your code, since my code is too slow and I cannot wait to find out the reason.
👍

**Comment by LetsGoFir** [comment]
@pengsida 
Hi~ I want to know the performance you have released named 197.pth on cityscapes, since I tested it and got 36.8 on val dataset. Is it normal?

**Comment by pengsida** [comment]
Yes, it is normal.
The coco evaluator gives lower results than the cityscapes official evaluator.
```
# use coco evaluator
python run.py --type evaluate --cfg_file configs/city_rcnn_snake.yaml
# use the cityscapes offical evaluator
python run.py --type evaluate --cfg_file configs/city_rcnn_snake.yaml test.dataset CityscapesVal
```

**Comment by LetsGoFir** [comment]
@pengsida 
It would be nice if you could have some comments about
'inp', 'meta', 'act_hm', 'awh', 'act_ind', 'act_01', 'ct_01', 'cp_hm', 'cp_wh', 'cp_ind', 'cp_01', 'i_it_4py', 'c_it_4py', 'i_gt_4py', 'c_gt_4py', 'i_it_py', 'c_it_py', 'i_gt_py', 'c_gt_py'

Although I have known some words like 'cp' means  component, etc.
: )

**Comment by pengsida** [comment]
I have updated the code https://github.com/zju3dv/snake/commit/77a6b49dbe93a5415aed28c23e5efd165e62ca59

**Comment by LetsGoFir** [comment]
> I have updated the code [77a6b49](https://github.com/zju3dv/snake/commit/77a6b49dbe93a5415aed28c23e5efd165e62ca59)

Thanks!

**Comment by fornote** [comment]
> I have updated the code [77a6b49](https://github.com/zju3dv/snake/commit/77a6b49dbe93a5415aed28c23e5efd165e62ca59)

hi, there is no Explanation about 'ct_01', and I want to know about it.

**Comment by pengsida** [comment]
It indicates if there is en element or not: https://github.com/zju3dv/snake/blob/master/lib/datasets/collate_batch.py#L24

**Comment by fornote** [comment]
> It indicates if there is en element or not: https://github.com/zju3dv/snake/blob/master/lib/datasets/collate_batch.py#L24

thank u very much.

**Comment by michaeltrs** [comment]
Congratulations for your wonderful work! 
Could you emphasize the key details between the graph convolution layer of curve-gcn and the circular convolution layers? From what I understand the key difference is that graph convolution utilises fewer parameters for the same number of adjacent vetrices, a set of weights for the center node and a set of weights applied to all nodes in the neighbourhood, while circular convolution uses a different set of weights for each neighbour.

**Comment by pengsida** [comment]
Yes, you are right.
![image](https://user-images.githubusercontent.com/11582122/148022720-d62d0a27-7a55-431c-8af4-fd0db4ba705f.png)

**Comment by michaeltrs** [comment]
thank you!
